Make window.open use correct process when called from a site isolated iframe

achristensen07 opened the pull request:
#### eab78766e024f6411cb60e10a3ed18b84e84aa30
<pre>
Make window.open use correct process when called from a site isolated iframe
<a href="https://bugs.webkit.org/show_bug.cgi?id=271514">https://bugs.webkit.org/show_bug.cgi?id=271514</a>

Reviewed by Charlie Wolfe.

When JS calls window.open, WebPageProxy::createNewPage gives an API::PageConfiguration
(which is now equivalent to a WKWebViewConfiguration) to the UIClient containing state
that must be used to make the new WKWebView.  Before site isolation, that state was just
the related page, whose process must be shared to keep functionality working such as
iframes with the same origin as the opener being able to directly access the opener&apos;s DOM.
With site isolation, the state that is needed to keep this functionality working is the
BrowsingContextGroup, which contains the maps of domains to processes, and the opener
frame identifier, which needs to be used to put the opened frame initially in the same
process as the opener while the opened frame contains only about:blank.  From about:blank
it can then use the BrowsingContextGroup to navigate to the correct process for any domain.

Unfortunately, though, there is SPI to do tricky things with the related page, and
WebKitTestRunner uses this SPI to retrofit an unrelated WKWebViewConfiguration into looking
enough like the suggested WKWebViewConfiguration that WebKit accepts it and uses it correctly.
In order to keep this working, I needed to make additional SPI to copy the site isolation
state from one WKPageConfigurationRef to another.  Conveninetly, WKPageConfigurationRef
can now be converted to WKWebViewConfiguration using toll-free-bridging.

It looks like all that work I did over the last week to clean up WKWebViewConfiguration
is already paying off!

I verified this fixes http/tests/site-isolation/iframe-and-window-open.html
but only how we used to run it before 276416@main.  Additional work is being done
on our strategy for running existing layout tests to verify site isolation correctness.

* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::setBrowsingContextGroup):
(API::PageConfiguration::openerFrameID const):
(API::PageConfiguration::setOpenerFrameID):
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
* Source/WebKit/UIProcess/API/C/WKPageConfigurationRef.cpp:
(WKPageConfigurationCopySiteIsolationState):
* Source/WebKit/UIProcess/API/C/WKPageConfigurationRef.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::createNewPage):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createWebPage):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::platformCreateOtherPage):

Canonical link: <a href="https://commits.webkit.org/276636@main">https://commits.webkit.org/276636@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/93077f0c29d2bcea8bc382c463c641ea31345279

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45235 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24353 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47755 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47898 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41243 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47542 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28541 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21750 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37101 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45813 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21403 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39014 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18199 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18816 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40098 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3282 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41499 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40414 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49603 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20217 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16748 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44135 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21523 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42934 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6288 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21211 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7447,6 +7447,8 @@ void WebPageProxy::createNewPage(WindowFeatures&& windowFeatures, NavigationActi
     Ref configuration = this->configuration().copy();
     configuration->setRelatedPage(*this);
+    configuration->setBrowsingContextGroup(m_browsingContextGroup);
+    configuration->setOpenerFrameID(originatingFrameInfoData.frameID);
```
Does `configuration().copy()` not copy `openerFrameID` and `browsingContextGroup`?

achristensen07 replied:
It does, but the openerFrameID was nullopt and now we need it to be set.  So the first of these two lines isn't needed but the second is.

szewai commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -1223,8 +1223,12 @@ Ref<WebPageProxy> WebProcessPool::createWebPage(PageClient& pageClient, Ref<API:
+    auto openerFrameID = pageConfiguration->openerFrameID();
+    if (pageConfiguration->preferences().siteIsolationEnabled() && openerFrameID) {
+        if (RefPtr openerFrame = WebFrameProxy::webFrame(*openerFrameID))
+            process = &openerFrame->process();
```
`process` is not set when `!openerFrame`, which will fail the assertion `ASSERT(process);` below.

charliewolfe replied:
I think that is ok. `openerFrame` should always exist if `openerFrameID` does.

szewai replied:
I don't think that's correct assumption because this function is invoked with configuration object passed by client, and client can technically set `openerFrameID` to anything. 
Even client uses the same configuration object created by WebKit, they can create the WebView asynchronously, so the frame / page may not exist when client invokes the function (see `UIDelegate::UIClient::createNewPage`).

If we are sure `openerFrame` exists, this should be written without the if statement for clarity:
```
RefPtr openerFrame = WebFrameProxy::webFrame(*openerFrameID)
RELEASE_ASSERT(openerFrame);
process = &openerFrame->process();
```

charliewolfe replied:
I see. Then yeah that assumption is wrong.

achristensen07 replied:
decidePolicyForNavigationAction is async, so the opener could actually be destroyed before this, so we don't want to RELEASE_ASSERT.  I'll make it fall back to the other process selection code if that happens.

achristensen07 replied:
The frame may also have navigated to a different origin.  I'm going to remember the process instead of the frame.

charliewolfe commented on Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
```diff
@@ -100,6 +100,21 @@ BrowsingContextGroup& PageConfiguration::browsingContextGroup() const
     return m_data.browsingContextGroup.get();
 }
 
+void PageConfiguration::setBrowsingContextGroup(BrowsingContextGroup& group)
```
Can this take a `RefPtr<>&&` like the other ones in this file?

achristensen07 replied:
Sure

achristensen07 committed:
Make window.open use correct process when called from a site isolated iframe
https://bugs.webkit.org/show_bug.cgi?id=271514

Reviewed by Charlie Wolfe.

When JS calls window.open, WebPageProxy::createNewPage gives an API::PageConfiguration
(which is now equivalent to a WKWebViewConfiguration) to the UIClient containing state
that must be used to make the new WKWebView.  Before site isolation, that state was just
the related page, whose process must be shared to keep functionality working such as
iframes with the same origin as the opener being able to directly access the opener's DOM.
With site isolation, the state that is needed to keep this functionality working is the
BrowsingContextGroup, which contains the maps of domains to processes, and the opener
frame identifier, which needs to be used to put the opened frame initially in the same
process as the opener while the opened frame contains only about:blank.  From about:blank
it can then use the BrowsingContextGroup to navigate to the correct process for any domain.

Unfortunately, though, there is SPI to do tricky things with the related page, and
WebKitTestRunner uses this SPI to retrofit an unrelated WKWebViewConfiguration into looking
enough like the suggested WKWebViewConfiguration that WebKit accepts it and uses it correctly.
In order to keep this working, I needed to make additional SPI to copy the site isolation
state from one WKPageConfigurationRef to another.  Conveninetly, WKPageConfigurationRef
can now be converted to WKWebViewConfiguration using toll-free-bridging.

It looks like all that work I did over the last week to clean up WKWebViewConfiguration
is already paying off!

I verified this fixes http/tests/site-isolation/iframe-and-window-open.html
but only how we used to run it before 276416@main.  Additional work is being done
on our strategy for running existing layout tests to verify site isolation correctness.

* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::setBrowsingContextGroup):
(API::PageConfiguration::openerFrameID const):
(API::PageConfiguration::setOpenerFrameID):
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
* Source/WebKit/UIProcess/API/C/WKPageConfigurationRef.cpp:
(WKPageConfigurationCopySiteIsolationState):
* Source/WebKit/UIProcess/API/C/WKPageConfigurationRef.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::createNewPage):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createWebPage):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::platformCreateOtherPage):

Canonical link: https://commits.webkit.org/276636@main
